pkg/maps/ctmap/ctmap.go

```diff
@@ -510,6 +517,108 @@ func GC(m *Map, filter *GCFilter) int {
 	return doGC(m, filter)
 }
 
+// PurgeOrphanNATEntries removes orphan SNAT entries. We call an SNAT entry
+// orphan if it does not have either a corresponding CT entry or an SNAT entry
+// in reverse order.
+//
+// The former case can happen when the CT entry is removed by the LRU eviction
+// which happens when the CT map becomes full. The latter case can happen for
+// the same LRU eviction reasons.
+//
+// PurgeOrphanNATEntries() is triggered by the datapath via the GC signaling
+// mechanism. When the datapath SNAT fails to find free mapping after
+// SNAT_SIGNAL_THRES attempts, it sends the signal via the perf ring buffer.
+// The consumer of the buffer invokes the function.
+//
+// The SNAT is being used for the following cases:
+// 1. By NodePort BPF on an intermediate node before fwd'ing request from outside
+//     to a destination node.
+// 2. A packet from local endpoint sent to outside (BPF-masq).
+// 3. A packet from a host local application (i.e. running in the host netns)
+//    This is needed to prevent SNAT from hijacking such connections.
+// 4. By DSR on a backend node to SNAT responses with service IP+port before
+//    sending to a client.
+//
+// In the case of 1-3, we always create a CT_EGRESS CT entry. This allows the
+// CT GC to remove corresponding SNAT entries. In the case of 4, will create
+// CT_INGRESS CT entry. See the unit test TestOrphanNatGC for more examples.
+//
+// The function only handles 1-3 cases, the 4. case is TODO(brb).
+func PurgeOrphanNATEntries(ctMap *Map) *NatGCStats {
+	if option.Config.NodePortMode == option.NodePortModeDSR ||
+		option.Config.NodePortMode == option.NodePortModeHybrid {
+		return nil
+	}
+
+	natMap := mapInfo[ctMap.mapType].natMap
+	if natMap == nil {
+		return nil
+	}
+	stats := newNatGCStats(natMap)
+
+	cb := func(key bpf.MapKey, value bpf.MapValue) {
+		natKey := key.(*nat.NatKey4)
+		natVal := value.(*nat.NatEntry4)
+
+		if natKey.GetFlags()&tuple.TUPLE_F_IN == 1 { // natKey is r(everse)tuple
+			ctKey := &tuple.TupleKey4Global{tuple.TupleKey4{
+				// Workaround #5848
+				SourceAddr: natKey.SourceAddr,
+				DestPort:   natKey.SourcePort,
+				DestAddr:   natVal.Addr,
+				SourcePort: natVal.Port,
+				NextHeader: natKey.NextHeader,
+				Flags:      tuple.TUPLE_F_OUT,
+			}}
+			if _, err := ctMap.Lookup(ctKey); err != nil && errors.Is(err, unix.ENOENT) {
```

Comment:
I'm curious, do we need the != nil check?
```suggestion
			if _, err := ctMap.Lookup(&ctKey); errors.Is(err, unix.ENOENT) {
```

Reply:
(Completely understandable if we're trying to avoid reflection in `errors.Is()` or something)